### PR TITLE
Bitmap join optimization

### PR DESCRIPTION
**Pull request by templarzq**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Performance Improvement
   improve bitmap involved join performance 

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

...

Detailed description / Documentation draft:
  during the join , bitmap
 table1: 
dimCol                                             
       -------                                            
              1                                                   
              1
              1
              .....

table2:
      
dimCol    bitmapCol
---------------------------  
    1               large bitmap ( e.g.  contains 30million numbers)

when table1 join tabl2  on dimCol,   the large bitmap will be copied for N times,  N is table1's   same value row count.
so clickhouse will spend much time on bitmap create and destroy,the query execution will be very slow.

so I modified the code of class AddedColumns in HashJoin.cpp  like below to improve this process:  
```
template <bool has_defaults>
void appendFromBlock(const Block & block, size_t row_num)
{
       if constexpr (has_defaults)
            applyLazyDefaults();
        std::unordered_map<Block*, std::unordered_map<size_t,size_t> >::iterator iter;
        std::unordered_map<size_t,size_t>::iterator iter2;
        bool bFound = false;
        size_t pos = 0;
        for (size_t j = 0; j < right_indexes.size(); ++j){
            if (std::string(columns[j]->getFamilyName())!="AggregateFunction"){  //not aggFuncCol,copy data directly
                columns[j]->insertFrom(*block.getByPosition(right_indexes[j]).column, row_num);
            }else{ // copy pointer for same bitmap
              if(!bFound){
                    iter = copiedDataInfo.find((Block*)&block);
                    if(iter!=copiedDataInfo.end()){    //found block
                        iter2 = iter->second.find(row_num);
                        if(iter2!=iter->second.end()){  //found row
                            pos = iter2->first;
                            auto pCol = typeid_cast<ColumnAggregateFunction *>(columns[j].get());
                            pCol->getData().push_back(pCol->getData()[pos]);
                        }else{
                            pos = columns[j]->size();
                            iter->second.insert(std::pair<size_t,size_t>(row_num,pos));
                            columns[j]->insertFrom(*block.getByPosition(right_indexes[j]).column, row_num);
                        }
                    }else{
                        pos = columns[j]->size();
                        std::unordered_map<size_t,size_t> rowInfo;
                        rowInfo.insert(std::pair<size_t,size_t>(row_num,pos));
                        copiedDataInfo.insert(std::pair<Block*, std::unordered_map<size_t,size_t> >((Block*)&block,
                            rowInfo));
                        columns[j]->insertFrom(*block.getByPosition(right_indexes[j]).column, row_num);
                    }
                    bFound = true;
            }else{
                auto pCol = typeid_cast<ColumnAggregateFunction *>(columns[j].get());
                pCol->getData().push_back(pCol->getData()[pos]);
            }
          }
        }
    }
```

.......

add definition of info container in class AddedColumns:
```
private:
   ....
   std::unordered_map<Block*, std::unordered_map<size_t,size_t> > copiedDataInfo;
```

and we should modify ~ColumnAggregateFunction to destroy bitmap  data correctly :
```
ColumnAggregateFunction::~ColumnAggregateFunction()
{
    if (!func->hasTrivialDestructor() && !src){
    //    for (auto * val : data)
    //
        for(size_t i=0;i<data.size();++i){
            auto  val = data[i];
            if(val==NULL){
                continue;
            }

            for(size_t j=i;j<data.size();++j){
                if(data[j]==val){
                    data[j]= NULL;
                }
            }

            func->destroy(val);
        }
    }
}
```

### CODE REVIEW

**Comment by templarzq** [comment]
test case :
drop table if EXISTS test_bm2;

drop table if EXISTS test_bm_join2;

create table test_bm2(
	dim UInt64,
	id  UInt64) 
ENGINE = MergeTree()
ORDER BY( dim )
SETTINGS index_granularity = 8192;

create table test_bm_join2( 
  dim UInt64,
  ids  AggregateFunction(groupBitmap, UInt64) )
ENGINE = MergeTree()
ORDER BY(dim)
SETTINGS index_granularity = 8192;

truncate table test_bm

insert into test_bm2 SELECT 1,`number` FROM numbers(0, 3000)

select * from test_bm2

truncate table test_bm_join2

insert into test_bm_join2 SELECT 1, bitmapBuild(range(toUInt64(0),toUInt64(30000000)))

select * from test_bm_join2


select a.dim,bitmapCardinality(b.ids)  from test_bm2 a left join test_bm_join2 b using(dim)

**Comment by templarzq** [comment]
just reverse the join order,the query speed will change very much:
select a.dim,bitmapCardinality(b.ids)  from test_bm_join2  b right join test_bm2 a  using(dim)

**Comment by l1t1** [comment]
please use ``` instead of ... to display code in good format.

**Comment by templarzq** [comment]
@l1t1  ok,the code formatted.

**Comment by templarzq** [comment]
this draft can only optimize bitmap copy an join action during the same block. if there are same bitmap in different blocks,this draft has no optimization. On the other way , search for same bitmap on every row  will cost some time ,this will slow the query without duplicated bitmap  somehow.

**Comment by alexey-milovidov** [comment]
Maybe revive this PR?

File: src/Interpreters/HashJoin.cpp [hunk 1]
```diff
@@ -726,6 +759,7 @@ class AddedColumns
     std::optional<TypeIndex> asof_type;
     ASOF::Inequality asof_inequality;
     const IColumn * left_asof_key = nullptr;
+    std::unordered_map<Block*, std::unordered_map<size_t,size_t> > copiedDataInfo;
```
Comment (alexey-milovidov) [inline src/Interpreters/HashJoin.cpp:762]: Missing a comment. It is difficult to understand, what is "copied data info"?

File: src/Interpreters/HashJoin.cpp [hunk 2]
```diff
@@ -687,9 +687,42 @@ class AddedColumns
     {
         if constexpr (has_defaults)
             applyLazyDefaults();
-
-        for (size_t j = 0; j < right_indexes.size(); ++j)
-            columns[j]->insertFrom(*block.getByPosition(right_indexes[j]).column, row_num);
+	std::unordered_map<Block*, std::unordered_map<size_t,size_t> >::iterator iter;
+        std::unordered_map<size_t,size_t>::iterator iter2;
+        bool bFound = false;
```
Comment (alexey-milovidov) [inline src/Interpreters/HashJoin.cpp:692]: What is `b` in `bFound`?

File: src/Interpreters/HashJoin.cpp [hunk 3]
```diff
@@ -687,9 +687,42 @@ class AddedColumns
...
+        size_t pos = 0;
+        for (size_t j = 0; j < right_indexes.size(); ++j){
+            if (std::string(columns[j]->getFamilyName())!="AggregateFunction"){  //not aggFuncCol,copy data directly
```
Comment (alexey-milovidov) [inline src/Interpreters/HashJoin.cpp:695]: This string comparison does not look alright.

**Comment by alexey-milovidov** [comment]
Please resubmit.

**Pull request closed without merging**